fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,577 @@
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final int MAX_OPEN_FILES = Integer.MAX_VALUE;
+  // Limits the number of translated (FUSE to TachyonURI)  paths that are kept
+  // in memory
+  private static final int MAX_CACHED_PATHS = 500;
+  private static final long[] UID_AND_GID = TachyonFuseUtils.getUidAndGid();
+
+  private final TachyonFileSystem mTFS;
+  private final Path mMountPoint;
+  private final Path mTachyonRootPath;
+  private final String mTachyonMaster;
+  // Keeps a cache of the most recently translated paths from String to TachyonURI
+  private final LoadingCache<String, TachyonURI> mPathResolverCache;
+
+  private final Object mOpenFilesLock;
+  // Table of open files with corresponding InputStreams and OutptuStreams
+  private final Map<Long, OpenFileEntry> mOpenFiles;
+  private long mOpenFileIds;
+
+  public TachyonFuseFs(TachyonFuseOptions opts) {
+    super();
+    mTFS = TachyonFileSystem.TachyonFileSystemFactory.get();
+    mMountPoint = Paths.get(opts.getMountPoint());
+    mTachyonMaster = opts.getMasterAddress();
+    mTachyonRootPath = Paths.get(opts.getTachyonRoot());
+    mOpenFileIds = 0L;
+    mOpenFiles = new HashMap<Long, OpenFileEntry>();
+    mOpenFilesLock = new Object();
+
+    mPathResolverCache = CacheBuilder.newBuilder()
+        .maximumSize(MAX_CACHED_PATHS)
+        .build(new PathCacheLoader());
+
+    Preconditions.checkArgument(mMountPoint.isAbsolute(),
+        "mount point should be an absolute path");
+    Preconditions.checkArgument(mTachyonRootPath.isAbsolute(),
+        "tachyon root path should be absolute");
+  }
+
+  @Override
+  public int create(String path, @mode_t long mode, FuseFileInfo fi) {
+    int ret = 0;
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    final int flags = fi.flags.get();
+    LOG.trace("create({}, {}) [Tachyon: {}]", path, Integer.toHexString(flags), turi);
+    LOG.warn("{}: mode is ignored in tachyon-fuse", path);
+    if ((flags & 3) != O_WRONLY.intValue()) {
+      LOG.error("Files can only be created in O_WRONLY mode ({})", path);
+      return -ErrorCodes.EACCES();
+    }
+
+    try {
+
+      final OpenFileEntry ofe = new OpenFileEntry();
+      ofe.mIn = null;
+      ofe.mOut = mTFS.getOutStream(turi);
+      LOG.debug("Tachyon OutStream created for {}", path);
+
+      synchronized (mOpenFilesLock) {
+        if (mOpenFiles.size() == MAX_OPEN_FILES) {
+          LOG.error("Cannot open {}: too many open files", turi);
+          return ErrorCodes.EMFILE();
+        }
+        mOpenFiles.put(mOpenFileIds, ofe);
+        fi.fh.set(mOpenFileIds);
+
+        // Assuming I will never wrap around (2^64 open files are quite a lot anyway)
+        mOpenFileIds += 1;
+      }
+      LOG.debug("{} created and opened in O_WRONLY mode", path);
+
+    } catch (FileAlreadyExistsException e) {
+      LOG.debug("File {} already exists", turi, e);
+      ret = ErrorCodes.EEXIST();
```

Comment:
Can we `return -ErrorCodes.EEXIST()` here and in the other exceptions, then return 0 if we make it to the end to indicate success? Then we don't need the `ret` variable any more, and its consistent with the other two places were we return early in this method.